EDS API: Additional Headers to enable EBSCO to assist with BOT protection

cwolfebsco opened the pull request:
Due to the nature of EDS API integration, EBSCO's web application firewall (WAF) does not have sufficient data to distinguish bots from users. By sending some end-user data to EDS API, EBSCO's WAF can make more 
informed decisions regarding bots. 

This is important to guarantee accurate COUNTER 5 usage reports. The Counter Code of Practice 5.1 stipulates:
"Activity generated by internet robots and crawlers MUST be excluded from all COUNTER usage reports."  Please also see https://cop5.countermetrics.org/en/5.1/07-processing/08-internet-robots-and-crawlers.html

If a site is already protecting their site from bots, e.g. through your own WAF, they might not wish to enable this.

Please note that this will send the end-user's IP Address & User Agent as x-eis-enduser-ip-address
x-eis-enduser-user-agent and sites might wish to consult their privacy officer.

This Pull Request:

* adds new section AdditionalHeaders to EDS.ini
* uses \VuFind\Net\UserIpReader's getUserIp function for IP to report x-eis-enduser-ip-address
* will attempt to parse config.ini > Site > generator or use client provided values

cwolfebsco committed:
Addition to EDS.ini

Potential new values:
* send_user_ip
* report_vendor
* report_vendor_version

We could considered to try and read config.ini -> Site -> generator and split at a space instead

cwolfebsco committed:
add option values into constructor

define:
* send_user_ip

If that is true also set:
* reportvendor
* report_vendor_version

cwolfebsco committed:
include x-eis IP & Vendor related headers

to figure out how to use use VuFind\Net\UserIpReader in here, which would be better than trying to build my own

cwolfebsco committed:
pass ip to report as an option

I found it easier to add ip to report in the EDS BackendFactory - but there must be a more appropriate way.

Question: I am using getUserIp() as it will mean there is only a single place to maintain this and it will be consistent

cwolfebsco committed:
Fix code layout issues, based on phpcs-console recommendation

cwolfebsco committed:
Code Improvements

as actioned by by vendor/bin/phing php-cs-fixer

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -443,3 +443,20 @@ DetailPageFormat = 'Long'
+
+; send_user_ip = true
+; report_vendor = "VuFind"
+; report_vendor_version = "10.1"
```
I would suggest using [VuFind\Config\Version](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Config/Version.php) instead of relying on configuration. That way it will be much harder for people to accidentally misreport their version.

Also, a tiny nitpick: I'd recommend adding a line break to the end of this line (we try to end all files with a line break).

otzbergnet replied:
I had thought of some strange logic to parse the "generator" from config.ini if report_vendor & report_vendor_version are not set. I propose to use the VuFind\Config\Version and remove report_vendor and report_vendor_version and simply hard code VuFind

demiankatz replied:
Yes, that seems simplest!

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -494,6 +534,18 @@ protected function call(
         }
+        if ($this->sendUserIp) {
+            $headers['x-eis-enduser-ip-address'] = $this->ipToReport ?? '-';
+            $headers['x-eis-enduser-user-agent'] = $_SERVER['HTTP_USER_AGENT']
```
I'd recommend injecting this via the factory instead of accessing a superglobal directly in the class. This will make testing easier, among other things. And in the factory, I'd recommend accessing the server variables through the `Request` service in the container, instead of directly via superglobal.

otzbergnet replied:
Just to be sure, you are suggesting I should get to HTTP_USER_AGENT by going through the getServer() method of \VuFind\Http\PhpEnvironment\Request?

demiankatz replied:
Exactly, yes!

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco -- just a couple of suggestions that I think should be helpful. As always, let me know if more details are needed.

dltj commented on config/vufind/EDS.ini:
```diff
@@ -443,3 +443,20 @@ DetailPageFormat = 'Long'
+; If you are already protecting your site from bots, e.g. through your own WAF, you might not wish to enable this
+
+; Please note that this will send the end-user's IP Address & User Agent as and x-eis-enduser-ip-address
+; x-eis-enduser-user-agent and you might thus wish to consult your privacy officer.
```
No suggestions from me other than correcting a typo ("as an" rather than "as and"). Thank you for the extra care in explaining what is being forwarded on through the API.

otzbergnet replied:
You can always count on my silly mistakes :) Thank you for pointing it out and sorry for having missed it

demiankatz replied:
I think there may still be some words missing here -- should it actually read something like this?

```
; Please note that this will send the end-user's IP Address & User Agent to EBSCO as x-eis-enduser-ip-address
; and x-eis-enduser-user-agent headers, so you might thus wish to consult your privacy officer.
```

cwolfebsco committed:
Addressing Feedback from Reviewers

Thank you for reviewing this PR and making it better. So much to learn, so little knowledge at my end :) Happy Thanksgiving!

cwolfebsco committed:
fix: forget to define property :(

cwolfebsco committed:
fix: spaces issue

cwolfebsco committed:
spaces are the worst :)

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -494,6 +534,14 @@ protected function call(
+            if (!empty($this->reportVendorVersion)) {
+                $headers['x-eis-vendor-version'] = $this->reportVendorVersion;
+            }
+        }
```
It might make sense to refactor all of the code that builds the `$headers` array to a separate `protected function getRequestHeaders(array $headerParams = []): array`. That way, if somebody wants to customize something like `x-eis-vendor` they could easily do so through a child class without having to rewrite the whole call() function. It makes things more customizable without requiring you to worry about extra config settings, etc.

Also, a related note: the function documentation here suggests that `$headerParams` must be an array, but the code supports it being null. I'm not sure where this inconsistency came from, but it's something we probably ought to address at some point. Not sure if it's worth doing as part of this work, but for now, I'd suggest only accepting an array in `getRequestHeaders` and then calling it here with `$this->getRequestHeaders($headerParams ?? [])`. That way, you can eliminate the null check in the new support function (since iterating an empty array seems simpler than adding a null check) but still maintain backward compatibility with the existing behavior.

otzbergnet replied:
not sure how to do it yet, but I like that idea :) Thank you for your patience with my Pull Requests and sorry for ruining your vacation...

demiankatz replied:
If you'd like me to open a separate PR to do the header refactoring so that you can then merge the changes into here, I can do that. And aren't you supposed to be on vacation too? ;-)

otzbergnet replied:
Half the fun to learn something new :) You might end up having to help me with pointers, but if you are ok with it, I'd like to try it.

You might have noticed that I am posting from my private account, because I am indeed on vacation. But jet lag is terrible and this gives me something useful to do before the sun comes up, and even at work this is not rally my "job", but I enjoy doing it - and the community...

demiankatz replied:
Yes, please feel free, and I'm happy to help if you have questions along the way!

cwolfebsco committed:
fix: my terrible grammar

cwolfebsco committed:
Attempt: create getRequestHeaders

based on feedback introduced a getRequestHeaders function, that will build out the headers.

Two calls in base.php had null in the call to the function, replaced those with empty arrays

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -506,6 +532,41 @@ protected function call(
+    /**
+     * Creat Header Array for Call Function
+     *
+     * @param array $headerParams an array (could be empty) of headders to build
```
Typo fix:

```suggestion
     * @param array $headerParams An array (could be empty) of headers to build
```

demiankatz reviewed (changes requested):
Your refactoring looks great! I'm just requesting a minor typo fix, but I'll go ahead and merge the fix myself so you don't have to do any more work... :-)

demiankatz committed:
Update module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php

demiankatz commented on module/VuFind/src/VuFind/Search/Factory/EdsBackendFactory.php:
```diff
@@ -103,6 +110,8 @@ public function __invoke(ContainerInterface $sm, $name, array $options = null)
         $this->setup($sm);
         $this->edsConfig = $this->getService(\VuFind\Config\PluginManager::class)
             ->get($this->getServiceName());
+        $this->vuFindConfig = $this->getService(\VuFind\Config\PluginManager::class)
```
This is no longer used, so I think you can remove the assignment and property from the factory.

demiankatz reviewed (changes requested):
Oops, noticed one more bit of cleanup that needs to be done (I think it's a leftover from the removed generator parsing logic).

I haven't done hands-on testing yet, but based on an eyeball read, the rest of this looks great now!

cwolfebsco committed:
Remove $vuFindConfig

This was a left over from a previous approach and I missed removing it

cwolfebsco committed:
Merge branch 'EDS-API-Enable-EBSCO-to-support-bot-protection' of https://github.com/cwolfebsco/vufind into EDS-API-Enable-EBSCO-to-support-bot-protection

demiankatz reviewed (approved):
This looks good to me, @cwolfebsco, though of course I can't directly test that it works. Let me know if you need any more assistance before we merge this. (If you think it's ready to merge, the next step is to take it out of draft mode).

cwolfebsco commented:
@demiankatz  - thank you so much for working with me on this. My tests in a local VM have all worked the way I had intended, so I believe it to be ready to be merged. I appreciate the collaborative spirit and the positive reinforcement of best practices and suggestions to improve this.

demiankatz reviewed (approved):
Thanks, @cwolfebsco! I appreciate your collaboration as well; thanks for taking the time not to just to share your work but also to polish it. :-)
